Mathlib/MeasureTheory/Function/Intersectivity.lean

```diff
@@ -0,0 +1,130 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.MeasureTheory.Function.LpSpace
+import Mathlib.MeasureTheory.Integral.Average
+
+/-!
+# Bergelson's intersectivity lemma
+
+This file proves the Bergelson intersectivity lemma: In a finite measure space, a sequence of events
+that have measure at least `r` has an infinite subset whose finite intersections all have positive
+volume.
+
+This is in some sense a finitary version of the second Borel-Cantelli lemma.
+
+## References
+
+[Bergelson, *Sets of recurrence of `ℤᵐ`-actions and properties of sets of differences in
+`ℤᵐ`][bergelson1985]
+
+## TODO
+
+Restate the theorem using the upper density of a set of naturals, once we have it. This will make
+`strong_bergelson` be actually strong.
+
+Use the ergodic theorem to deduce the refinement of the Poincaré recurrence theorem proved by
+Bergelson.
+-/
+
+open Filter Function MeasureTheory Set
+open scoped BigOperators ENNReal NNReal
+
+variable {ι α : Type*} [MeasurableSpace α] {μ : Measure α} [IsFiniteMeasure μ] {r : ℝ≥0∞}
+
+/-- **Bergelson Intersectivity Lemma**: In a finite measure space, a sequence of events that have
+measure at least `r` has an infinite subset whose finite intersections all have positive volume.
+
+TODO: The infinity of `t` should be strengthened to `t` having positive natural density. -/
+lemma strong_bergelson {s : ℕ → Set α} (hs : ∀ n, MeasurableSet (s n)) (hr₀ : r ≠ 0)
```

Comment:
Am I correct that `strong_bergelson` and `weak_bergelson` only differs in the index? Perhaps you can call them `bergelson` and `bergelson'` for now.

Reply:
Sure